### PR TITLE
Add `forall` method to `JsResult` with `Option#forall` semantics

### PR DESCRIPTION
**Pull request by aaabramov**
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Typesafe CLA](https://www.typesafe.com/contribute/cla)?
* [x] Have you [squashed your commits](https://www.playframework.com/documentation/latest/WorkingWithGit#Squashing-commits)?
* [x] Have you added copyright headers to new files?
* [x] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?

## Purpose

Adds `forall` method to `JsResult` with `Option#forall` semantics

## References

Initial discussion: https://github.com/playframework/play-json/pull/178

### CODE REVIEW

File: play-json/shared/src/main/scala/play/api/libs/json/JsResult.scala [hunk 1]
```diff
@@ -221,6 +225,12 @@ sealed trait JsResult[+A] { self =>
   /** If this result is successful than check value with predicate '''p''', otherwise return '''false''' */
   def exists(p: A => Boolean): Boolean
 
+  /**
+   * If this result is successful than check value with predicate '''p''', otherwise return '''true'''.
+   * Follows [[Option.forall]] semantics
```
Comment (cchantep) [inline play-json/shared/src/main/scala/play/api/libs/json/JsResult.scala]: Option ~> Traversable

File: play-json/shared/src/test/scala/play/api/libs/json/JsResultSpec.scala [hunk 2]
```diff
@@ -77,4 +77,24 @@ class JsResultSpec extends WordSpec with MustMatchers {
 
   }
 
+  "JsSuccess#forall" should {
+
+    "return true for JsSuccess(x * 2).forall(_ % 2 == 0)" in {
+      JsSuccess(2).forall(_ % 2 == 0) mustBe true
```
Comment (cchantep) [inline play-json/shared/src/test/scala/play/api/libs/json/JsResultSpec.scala]: `mustBe <boolean>` vs `assert` in previous PR ... coherency

**Comment by aaabramov** [comment]
@cchantep PR updated.

File: play-json/shared/src/main/scala/play/api/libs/json/JsResult.scala [hunk 3]
```diff
@@ -71,6 +73,8 @@ case class JsError(errors: collection.Seq[(JsPath, collection.Seq[JsonValidation
 
   def exists(p: Nothing => Boolean): Boolean = false
 
+  def forall(p: Nothing => Boolean): Boolean = true
+
```
Comment (renatocaval) [inline play-json/shared/src/main/scala/play/api/libs/json/JsResult.scala]: I guess we must have it returning `true` just to align with other `forall` methods in Scala. However, this is something that always annoyed me. 

`forall` on the empty or error side always returns true in other libraries.

```scala
@ val optEmpty: Option[String] = None
optEmpty: Option[String] = None

@ optEmpty.forall(_.startsWith("a"))
res6: Boolean = true

@ optEmpty.forall(_.startsWith("b"))
res6: Boolean = true
```

In the above example, how does it even make any sense? If the option is empty, all predicates making assertions about its content should return false. The same is valid for Scala collections. 

We will have the same situation here.

Comment (renatocaval) [reply play-json/shared/src/main/scala/play/api/libs/json/JsResult.scala]: I asked about it in the Scala forum the motivations for this behaviour. 
https://users.scala-lang.org/t/forall-semantics/3303

Not that I believe we should change it. We should keep it consistent with other libraries.

Comment (aaabramov) [reply play-json/shared/src/main/scala/play/api/libs/json/JsResult.scala]: @renatocaval I realized that `_.forall` behavior is pretty lear after reading this blogpost: http://blog.originate.com/blog/2014/06/15/idiomatic-scala-your-options-do-not-match/

Specifically:
image
image

>It makes perfect sense in plain English: to say “all numbers are prime” is equivalent to saying “there does not exist a number that is not prime”.

Comment (renatocaval) [reply play-json/shared/src/main/scala/play/api/libs/json/JsResult.scala]: Thanks @aaabramov. 
I suspected that there was a solid reason for that. TIL
